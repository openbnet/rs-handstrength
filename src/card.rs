#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}