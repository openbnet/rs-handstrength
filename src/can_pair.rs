use crate::card::{Card, Suit};
use crate::sorting::*;
use itertools::Itertools;
use crate::can_libs::*;

pub fn can_have_pair(comcards: &Vec<Card>) -> CanHaveCombis {
    let mut can_have_combis: CanHaveCombis = Vec::new();
    let sorted = sort_cards(&replace_ace_as_high(&comcards));
    let sorted_values: Vec<u8> = sorted.iter().map(|s| s.value).collect();
    let grouped_cards = get_same_value_map(&sorted);
    let pair_groups: Vec<Vec<Card>> = grouped_cards.iter().filter(|g| g.len() == 2).cloned().collect();
    // println!("pair_groups {:?}", pair_groups);
    let all_values: Vec<u8> = (2..=14).rev().collect(); 
    if pair_groups.len() == 0 {
        for v in &all_values {
            if sorted_values.contains(&v) == false {
                // push user pair in hand here
                can_have_combis.push((
                    sorted.clone(),
                    vec![vec![vec![Card {value: *v, suit: Suit::A}, Card { value: *v, suit: Suit::A}]]]
                ))
            } 
            if sorted_values.contains(&v) == true {
                // User hits the board for a pair
                let kickers = all_values.iter()
                    .filter(|&kv| !sorted_values.contains(kv) && kv != *&v)
                    .cloned()
                    .collect::<Vec<u8>>();
    
                let hand_combinations = kickers.into_iter()
                    .map(|kicker| vec![Card { value: *v, suit: Suit::A }, Card {value: kicker, suit: Suit::A}])
                    .collect::<Vec<Vec<Card>>>();
    
                can_have_combis.push((sorted.clone(), vec![hand_combinations]));
            } 
    
        } 
    }
    else {
        // pair on board
        // user cant have pair in hand or else its two pair
        // user cant hit anything on board or its two pair
        let first_pair_value = pair_groups[0][0].value;
        let kickers = all_values.iter()
            .filter(|&kv| !sorted_values.contains(kv) )
            .cloned()
            .collect::<Vec<u8>>();
        // println!("kickers {:?}", &kickers);
        let hand_combinations = kickers.into_iter()
            .combinations(2)
            .map(|kicker| vec![vec![Card { value: kicker[0], suit: Suit::A }, Card {value: kicker[1], suit: Suit::A}]])
            .collect::<Vec<Vec<Vec<Card>>>>();

        can_have_combis.push((sorted.clone(), hand_combinations));
    }



    can_have_combis
}