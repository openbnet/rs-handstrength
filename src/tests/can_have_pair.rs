#[cfg(test)]
mod tests {
    use crate::card::{Card, Suit};
    use crate::can_pair::can_have_pair;
    use insta::assert_debug_snapshot;
    // #[test]
    fn nopair_board() {
        let cards = vec![
            Card { value: 1, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 10, suit: Suit::C },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
        ];
        let combis = can_have_pair(&cards);
        // println!("combis {:?}", combis);
        assert_debug_snapshot!(combis);
    }
    // #[test]
    fn pair1_board() {
        let cards = vec![
            Card { value: 1, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 10, suit: Suit::C },
            Card { value: 10, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
        ];
        let combis = can_have_pair(&cards);
        // println!("combis {:?}", combis);
        assert_debug_snapshot!(combis);
    }
    #[test]
    fn pair2_board() {
        let cards = vec![
            Card { value: 1, suit: Suit::H },
            Card { value: 8, suit: Suit::D },
            Card { value: 10, suit: Suit::C },
            Card { value: 10, suit: Suit::S },
            Card { value: 1, suit: Suit::S },
        ];
        let combis = can_have_pair(&cards);
        // println!("combis {:?}", combis);
        assert_debug_snapshot!(combis);
    }
}