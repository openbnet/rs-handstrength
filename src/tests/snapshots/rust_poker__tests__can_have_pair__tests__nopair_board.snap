---
source: src/tests/can_have_pair.rs
expression: combis
---
[
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 14,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 13,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 12,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 11,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 9,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 7,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 6,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 5,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 4,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 14,
                suit: H,
            },
            Card {
                value: 10,
                suit: C,
            },
            Card {
                value: 8,
                suit: D,
            },
            Card {
                value: 3,
                suit: S,
            },
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 13,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 12,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 11,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 7,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 6,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 5,
                        suit: A,
                    },
                ],
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 4,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
]
